sgkit_vcf/vcf_reader.py

```diff
@@ -0,0 +1,286 @@
+
+        variant_contig = np.empty(chunk_length, dtype="i1")
+        variant_position = np.empty(chunk_length, dtype="i4")
+        call_genotype = np.empty((chunk_length, n_sample, n_ploidy), dtype="i1")
+        call_genotype_phased = np.empty((chunk_length, n_sample), dtype=bool)
+
+        first_variants_chunk = True
+        for variants_chunk in chunks(region_filter(variants, region), chunk_length):
+
+            variant_ids = []
+            variant_alleles = []
+
+            for i, variant in enumerate(variants_chunk):
+                variant_id = variant.ID if variant.ID is not None else "."
+                variant_ids.append(variant_id)
+                max_variant_id_length = max(max_variant_id_length, len(variant_id))
+                variant_contig[i] = variant_contig_names.index(variant.CHROM)
+                variant_position[i] = variant.POS
+
+                alleles = [variant.REF] + variant.ALT
+                if len(alleles) > n_allele:
+                    alleles = alleles[:n_allele]
+                elif len(alleles) < n_allele:
+                    alleles = alleles + ([""] * (n_allele - len(alleles)))
+                variant_alleles.append(alleles)
+                max_variant_allele_length = max(
+                    max_variant_allele_length, max(len(x) for x in alleles)
+                )
+
+                gt = variant.genotype.array()
+                call_genotype[i] = gt[..., 0:-1]
+                call_genotype_phased[i] = gt[..., -1]
+
+            # Truncate np arrays (if last chunk is smaller than chunk_length)
+            if i + 1 < chunk_length:
+                variant_contig = variant_contig[: i + 1]
+                variant_position = variant_position[: i + 1]
+                call_genotype = call_genotype[: i + 1]
+                call_genotype_phased = call_genotype_phased[: i + 1]
+
+            variant_id = np.array(variant_ids, dtype="O")
+            variant_id_mask = variant_id == "."
+            variant_alleles = np.array(variant_alleles, dtype="O")
+
+            ds: xr.Dataset = create_genotype_call_dataset(
+                variant_contig_names=variant_contig_names,
+                variant_contig=variant_contig,
+                variant_position=variant_position,
+                variant_alleles=variant_alleles,
+                sample_id=sample_id,
+                call_genotype=call_genotype,
+                call_genotype_phased=call_genotype_phased,
+                variant_id=variant_id,
+            )
+            ds["variant_id_mask"] = (
+                [DIM_VARIANT],
+                variant_id_mask,
+            )
+            ds.attrs["max_variant_id_length"] = max_variant_id_length
+            ds.attrs["max_variant_allele_length"] = max_variant_allele_length
+
+            if first_variants_chunk:
+                # Enforce uniform chunks in the variants dimension
+                # Also chunk in the samples direction
+                encoding = dict(
```

Comment:
👍 

Being able to decouple the zarr chunking from the dask chunking like this is very cool